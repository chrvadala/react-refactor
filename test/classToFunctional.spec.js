const {patchString} = require("../src/stringUtils");
const {removeSpaces} = require("./testUtils");
const {classToFunctional} = require('../src/classToFunctional')
const {parse} = require('../src/parser')

const classTemplate = `
class ClassComp extends React.Component {
  constructor(props) {
    super(props);
  }
  doSomething(){
    //doSomething()
  }
  render() {
    let {def} = this.props
    let {props: {ghi}} = this
    return (
      <div>{this.props.abc}</div>
    );
  }
}
`.trim()

const functionalTemplate = `
function ClassComp(props){
  let self = {props};
  
  let {def} = self.props
  let {props: {ghi}} = self
  return (
    <div>{self.props.abc}</div>
  );
}
`

const genericClassTemplate = `
class ClassComp extends React.Component {
  constructor(props) {
    super(props);
  }
  doSomething(){
    //doSomething()
  }
}
`.trim()

describe('classToFunctional', () => {
  it('should convert class to functional comp', () => {
    let classComp = parse(classTemplate)
    let patch = classToFunctional(classTemplate, classComp.body[0])
    expect(Array.isArray(patch)).toBe(true)
    expect(patch).toMatchSnapshot()
    let output = patchString(classTemplate, patch)
    expect(removeSpaces(output)).toBe(removeSpaces(functionalTemplate))
  })
  it('should return false', () => {
    let classComp = parse(genericClassTemplate)
    let patch = classToFunctional(classTemplate, classComp.body[0])
    expect(patch).toBeFalsy()
  })

  describe('should supports different superclass', () => {
    [
      'React.PureComponent',
      'PureComponent',
      'Component',
    ].forEach(superClass => {
      it(`should support ${superClass}`, () => {
        let classCompVariant = classTemplate.replace('React.Component', superClass)
        let functionalCompVariant = functionalTemplate.replace('React.Component', superClass)
        let classComp = parse(classCompVariant)
        let patch = classToFunctional(classTemplate, classComp.body[0])
        expect(Array.isArray(patch)).toBe(true)
        expect(patch).toMatchSnapshot()
        let output = patchString(classCompVariant, patch)
        expect(removeSpaces(output)).toBe(removeSpaces(functionalCompVariant))
      })
    })
  })
})
