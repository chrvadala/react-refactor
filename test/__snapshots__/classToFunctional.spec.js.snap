// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`classToFunctional should convert class to functional comp 1`] = `
Array [
  Object {
    "operation": "insert",
    "payload": "function ClassComp(props)",
    "start": 0,
  },
  Object {
    "end": 98,
    "operation": "remove",
    "start": 0,
  },
  Object {
    "operation": "insert",
    "payload": "
	let self = {props};
",
    "start": 99,
  },
  Object {
    "operation": "insert",
    "payload": "self",
    "start": 116,
  },
  Object {
    "end": 120,
    "operation": "remove",
    "start": 116,
  },
  Object {
    "operation": "insert",
    "payload": "self",
    "start": 152,
  },
  Object {
    "end": 156,
    "operation": "remove",
    "start": 152,
  },
  Object {
    "operation": "insert",
    "payload": "self",
    "start": 182,
  },
  Object {
    "end": 186,
    "operation": "remove",
    "start": 182,
  },
]
`;
